### PR TITLE
Add support for tvOS [local tvOS WebRTC.framework required]

### PR DESCRIPTION
**Pull request by montanaeli**
Features	
* Added tvOS 16 support.

Fixed
* Added conditional statements on Capturer files, using If cases for tvOS.

Testing Environment
* Dependencies:
   * React 18.0.0
   * React Native tvOS 0.69.8-2
   * Local dependency for WebRTC.framework m112 build for tvOS.
* Simulators used:
   * tvOS 16.0, 16.1 and 16.3
   * iOS 16.4
   * Android Tiramisu
   * Android TV Tiramisu

For testing, it is necessary to add an Xcode reference to a local WebRTC.framework m112 (it is necessary to comment Jitsi dependency in the .podspec, to avoid this error:_`JitsiWebRTC (111.0.2)`, does not support `tvOS`_ on pod installation), also it must be used compatible versions of React and React Native tvOS.
The latest versions of React Native don’t support tvOS, that why it’s required to use react-native-tvos library instead.
Data channels have not been tested.

### CODE REVIEW

**Comment by 8BallBomBom** [comment]
Welcome and Thanks for the contribution 👍🏻

This looks very interesting, never figured we'd be supporting anything tv related.
Generally speaking though this looks good but will need a proper review and considerations.

Gotta say, not liking the look of the drop in replacement `react-native` package 🤔
The usual approach would be additional packages with additional functionality.
But it is what it is and understandable.

**Comment by montanaeli** [comment]
Here is the sample app we used to test it: [tvapp](https://github.com/millicast/streaming-sdk-react-native-getting-started/tree/tvapp).

Thanks in advance, we'll look forward to hearing from you 😄

**Comment by saghul** [comment]
I'll be reviewing this shortly!

**Comment by fcancela** [comment]
> Welcome and Thanks for the contribution 👍🏻
> 
> This looks very interesting, never figured we'd be supporting anything tv related. Generally speaking though this looks good but will need a proper review and considerations.
> 
> Gotta say, not liking the look of the drop in replacement `react-native` package 🤔 The usual approach would be additional packages with additional functionality. But it is what it is and understandable.

I'm glad you mentioned this. TBH, we didn't think to much on a way we could resolve this. So we are fully open to suggestions on what the right approach should be. 

Here's the official reference to this community mantained package.
https://reactnative.dev/docs/building-for-tv?tv=tvos
https://github.com/react-native-tvos/react-native-tvos

Just a note, It's actively being mantained and supports all platforms, it's not only for tvOS.

**Comment by 8BallBomBom** [comment]
Definitely had me stumped by that one specially when looking at the macOS and Windows packages.
They just give you an extra package and you're off to the races.

The right approach i'd say is what works and that applies better if the project is also kept up to date.
But they could definitely accomplish the same with a separate package which applies patches on top.

**Comment by saghul** [comment]
IIRC this is generally solved with Metro config so when packages depend on `react-native` it resolves to the tvOS flavour. See an example here using react-native-macos: https://github.com/react-native-webrtc/react-native-webrtc/blob/master/examples/GumTestApp_macOS/react-native.config.js

File: ios/RCTWebRTC/CaptureController.m [hunk 1]
```diff
@@ -1,3 +1,5 @@
+#if TARGET_OS_IOS
```
Comment (saghul) [inline ios/RCTWebRTC/CaptureController.m:1]: I believe this is also needed for macOS.

File: ios/RCTWebRTC/ScreenCapturer.m [hunk 2]
```diff
@@ -258,3 +260,5 @@ - (void)stream:(NSStream *)aStream handleEvent:(NSStreamEvent)eventCode {
 }
 
 @end
+
+#endif  // !TARGET_OS_OSX and !TARGET_OS_TVOS
```
Comment (saghul) [inline ios/RCTWebRTC/ScreenCapturer.m:264]: Isn't TARGET_OS_IOS more explicit here?

File: ios/RCTWebRTC/VideoCaptureController.h [hunk 3]
```diff
@@ -1,17 +1,23 @@
...
+    #import <WebRTC/RTCCameraVideoCapturer.h>
+#endif
 
 #import "CaptureController.h"
 
 @interface VideoCaptureController : CaptureController
-
-@property(nonatomic, readonly, strong) AVCaptureDeviceFormat *selectedFormat;
+#if !TARGET_OS_TV
+    @property(nonatomic, readonly, strong) AVCaptureDeviceFormat *selectedFormat;
```
Comment (saghul) [inline ios/RCTWebRTC/VideoCaptureController.h:12]: Note for future reference: https://developer.apple.com/documentation/avfoundation/avcapturedeviceformat is supported on tvOS 17.

File: ios/RCTWebRTC/WebRTCModule+Permissions.m [hunk 4]
```diff
@@ -29,27 +31,28 @@ - (AVMediaType)avMediaType:(NSString *)mediaType {
         reject(@"invalid_type", @"Invalid media type", nil);
         return;
     }
+        AVAuthorizationStatus status = [AVCaptureDevice authorizationStatusForMediaType:mediaType_];
```
Comment (saghul) [inline ios/RCTWebRTC/WebRTCModule+Permissions.m:34]: Why is this code re-indented? Please run `npm run format` which uses clang-format.

File: ios/react-native-webrtc-tvos/react_native_webrtc_tvos.m [hunk 5]
```diff
@@ -0,0 +1,10 @@
+//
```
Comment (saghul) [inline ios/react-native-webrtc-tvos/react_native_webrtc_tvos.m:1]: Ditto.

**Review by saghul — CHANGES_REQUESTED** [review]
Left some feedback, please take a look!

File: ios/RCTWebRTC/WebRTCModule+RTCMediaStream.m [hunk 6]
```diff
@@ -12,7 +15,9 @@
 #import "ScreenCaptureController.h"
 #import "ScreenCapturer.h"
 #import "TrackCapturerEventsEmitter.h"
-#import "VideoCaptureController.h"
+#if !TARGET_OS_TV
```
[the remark this answers is not in the document]
Comment (montanaeli) [reply ios/RCTWebRTC/WebRTCModule+RTCMediaStream.m:18]: Screenshot 2023-08-03 at 15 17 27

If I do it this way Xcode still cannot find AVFoundation capturer symbols, does this mean that we also have to define dummy methods inside those classes too?

[the remark this answers is not in the document]
Comment (montanaeli) [reply ios/RCTWebRTC/WebRTCModule+RTCMediaStream.m:18]:
```bash

Showing Recent Errors Only
/Users/emont/Desktop/RN-tvOS-SA/node_modules/react-native-webrtc/ios/RCTWebRTC/WebRTCModule+RTCMediaStream.m:83:17: 'AVCaptureDeviceFormat' is unavailable: not available on tvOS

/Users/emont/Desktop/RN-tvOS-SA/node_modules/react-native-webrtc/ios/RCTWebRTC/WebRTCModule+RTCMediaStream.m:83:53: Property 'selectedFormat' not found on object of type 'VideoCaptureController *'

/Users/emont/Desktop/RN-tvOS-SA/node_modules/react-native-webrtc/ios/RCTWebRTC/WebRTCModule+RTCMediaStream.m:229:13: 'AVCaptureDeviceFormat' is unavailable: not available on tvOS

/Users/emont/Desktop/RN-tvOS-SA/node_modules/react-native-webrtc/ios/RCTWebRTC/WebRTCModule+RTCMediaStream.m:229:49: Property 'selectedFormat' not found on object of type 'VideoCaptureController *'

/Users/emont/Desktop/RN-tvOS-SA/node_modules/react-native-webrtc/ios/RCTWebRTC/WebRTCModule+RTCMediaStream.m:253:5: 'AVCaptureDeviceDiscoverySession' is unavailable: not available on tvOS

/Users/emont/Desktop/RN-tvOS-SA/node_modules/react-native-webrtc/ios/RCTWebRTC/WebRTCModule+RTCMediaStream.m:254:77: 'AVCaptureDeviceTypeBuiltInWideAngleCamera' is unavailable: not available on tvOS

/Users/emont/Desktop/RN-tvOS-SA/node_modules/react-native-webrtc/ios/RCTWebRTC/WebRTCModule+RTCMediaStream.m:256:74: 'AVCaptureDevicePositionUnspecified' is unavailable: not available on tvOS

/Users/emont/Desktop/RN-tvOS-SA/node_modules/react-native-webrtc/ios/RCTWebRTC/WebRTCModule+RTCMediaStream.m:254:42: 'AVCaptureDeviceDiscoverySession' is unavailable: not available on tvOS

/Users/emont/Desktop/RN-tvOS-SA/node_modules/react-native-webrtc/ios/RCTWebRTC/WebRTCModule+RTCMediaStream.m:257:10: 'AVCaptureDevice' is unavailable: not available on tvOS

/Users/emont/Desktop/RN-tvOS-SA/node_modules/react-native-webrtc/ios/RCTWebRTC/WebRTCModule+RTCMediaStream.m:259:32: 'AVCaptureDevicePositionBack' is unavailable: not available on tvOS

/Users/emont/Desktop/RN-tvOS-SA/node_modules/react-native-webrtc/ios/RCTWebRTC/WebRTCModule+RTCMediaStream.m:261:39: 'AVCaptureDevicePositionFront' is unavailable: not available on tvOS

/Users/emont/Desktop/RN-tvOS-SA/node_modules/react-native-webrtc/ios/RCTWebRTC/WebRTCModule+RTCMediaStream.m:276:5: 'AVCaptureDeviceDiscoverySession' is unavailable: not available on tvOS

/Users/emont/Desktop/RN-tvOS-SA/node_modules/react-native-webrtc/ios/RCTWebRTC/WebRTCModule+RTCMediaStream.m:277:77: 'AVCaptureDeviceTypeBuiltInMicrophone' is unavailable: not available on tvOS

/Users/emont/Desktop/RN-tvOS-SA/node_modules/react-native-webrtc/ios/RCTWebRTC/WebRTCModule+RTCMediaStream.m:279:74: 'AVCaptureDevicePositionUnspecified' is unavailable: not available on tvOS

/Users/emont/Desktop/RN-tvOS-SA/node_modules/react-native-webrtc/ios/RCTWebRTC/WebRTCModule+RTCMediaStream.m:277:42: 'AVCaptureDeviceDiscoverySession' is unavailable: not available on tvOS

/Users/emont/Desktop/RN-tvOS-SA/node_modules/react-native-webrtc/ios/RCTWebRTC/WebRTCModule+RTCMediaStream.m:280:10: 'AVCaptureDevice' is unavailable: not available on tvOS

/Users/emont/Desktop/RN-tvOS-SA/node_modules/react-native-webrtc/ios/RCTWebRTC/WebRTCModule+RTCMediaStream.m:379:65: No visible @interface for 'VideoCaptureController' declares the selector 'switchCamera'


```

File: Documentation/tvOSInstallation.md [hunk 7]
```diff
@@ -0,0 +1,17 @@
+# tvOS Installation
```
[the remark this answers is not in the document]
Comment (montanaeli) [reply Documentation/tvOSInstallation.md:1]: We have some ideas we can discuss on Monday. We are also open to suggestions.

File: ios/RCTWebRTC.xcodeproj/project.pbxproj [hunk 8]
```diff
@@ -92,6 +94,15 @@
 /* End PBXFrameworksBuildPhase section */
 
 /* Begin PBXGroup section */
+		0A4A3E1F2A585B4900D3FECA /* react-native-webrtc-tvos */ = {
```
Comment (saghul) [inline ios/RCTWebRTC.xcodeproj/project.pbxproj:97]: Ditto.

File: ios/RCTWebRTC.xcodeproj/project.pbxproj [hunk 9]
```diff
@@ -92,6 +94,15 @@
...
+			isa = PBXGroup;
+			children = (
+				0A4A3E202A585B4900D3FECA /* react_native_webrtc_tvos.h */,
```
Comment (saghul) [inline ios/RCTWebRTC.xcodeproj/project.pbxproj:100]: Ditto.

File: ios/RCTWebRTC.xcodeproj/project.pbxproj [hunk 10]
```diff
@@ -105,6 +116,7 @@
 			isa = PBXGroup;
 			children = (
 				35A222311CB493C00015FD5C /* RCTWebRTC */,
+				0A4A3E1F2A585B4900D3FECA /* react-native-webrtc-tvos */,
```
Comment (saghul) [inline ios/RCTWebRTC.xcodeproj/project.pbxproj:119]: Ditto.

File: ios/RCTWebRTC.xcworkspace/xcshareddata/IDEWorkspaceChecks.plist [hunk 11]
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (saghul) [inline ios/RCTWebRTC.xcworkspace/xcshareddata/IDEWorkspaceChecks.plist:1]: I don't think this file is necesary.

File: ios/RCTWebRTC/WebRTCModule+Permissions.m [hunk 12]
```diff
@@ -45,11 +46,13 @@ - (AVMediaType)avMediaType:(NSString *)mediaType {
             break;
     }
 }
+#endif
 
+#if !TARGET_OS_TV
```
Comment (saghul) [inline ios/RCTWebRTC/WebRTCModule+Permissions.m:51]: Ditto.

File: ios/RCTWebRTC/WebRTCModule+RTCMediaStream.m [hunk 13]
```diff
@@ -46,12 +48,12 @@ - (RTCVideoTrack *)createVideoTrackWithCaptureController:
 
     return videoTrack;
 }
-
 /**
  * Initializes a new {@link RTCMediaTrack} with the given tracks.
  *
  * @return An array with the mediaStreamId in index 0, and track infos in index 1.
  */
+#if !TARGET_OS_TV
```
Comment (saghul) [inline ios/RCTWebRTC/WebRTCModule+RTCMediaStream.m:56]: Ditto.

File: ios/RCTWebRTC/WebRTCModule+RTCMediaStream.m [hunk 14]
```diff
@@ -137,6 +143,9 @@ - (RTCVideoTrack *)createScreenCaptureVideoTrack {
 }
 
 RCT_EXPORT_METHOD(getDisplayMedia : (RCTPromiseResolveBlock)resolve rejecter : (RCTPromiseRejectBlock)reject) {
+#if !TARGET_OS_TV
+    return nil
```
Comment (saghul) [inline ios/RCTWebRTC/WebRTCModule+RTCMediaStream.m:147]: You need to call reolve ot reject.

File: ios/RCTWebRTC/WebRTCModule+RTCMediaStream.m [hunk 15]
```diff
@@ -171,10 +180,12 @@ - (RTCVideoTrack *)createScreenCaptureVideoTrack {
  * if audio permission was not granted, there will be no "audio" key in
  * the constraints dictionary.
  */
+#if !TARGET_OS_TV
```
Comment (saghul) [inline ios/RCTWebRTC/WebRTCModule+RTCMediaStream.m:183]: Don't skip

File: ios/RCTWebRTC/WebRTCModule+RTCMediaStream.m [hunk 16]
```diff
@@ -346,6 +363,9 @@ - (RTCVideoTrack *)createScreenCaptureVideoTrack {
 }
 
 RCT_EXPORT_METHOD(mediaStreamTrackSetEnabled : (nonnull NSNumber *)pcId : (nonnull NSString *)trackID : (BOOL)enabled) {
+#if !TARGET_OS_TV
+    return nil
```
Comment (saghul) [inline ios/RCTWebRTC/WebRTCModule+RTCMediaStream.m:367]: Ditto, this function returns void.

**Review by saghul — CHANGES_REQUESTED** [review]
Left some more comments. Please remember to run `npm run format` too and commit the result.

File: ios/react-native-webrtc-tvos/react_native_webrtc_tvos.h [hunk 17]
```diff
@@ -0,0 +1,10 @@
+//
```
[the remark this answers is not in the document]
Comment (olbapbetan) [reply ios/react-native-webrtc-tvos/react_native_webrtc_tvos.h]: Seems to be an autogenerated file. We got rid of it.

**Comment by montanaeli** [comment]
Sorry about the delay, I hope I understood the changes you asked for properly

File: ios/RCTWebRTC/ScreenCaptureController.m [hunk 18]
```diff
@@ -1,3 +1,5 @@
+#if !TARGET_OS_OSX && !TARGET_OS_TV
```
Comment (saghul) [inline ios/RCTWebRTC/ScreenCaptureController.m:1]: Let's be explicit here and just say TARGET_OS_IOS

File: ios/RCTWebRTC/ScreenCapturePickerViewManager.m [hunk 19]
```diff
@@ -1,4 +1,4 @@
-#if !TARGET_OS_OSX
+#if !TARGET_OS_OSX && !TARGET_OS_TV
```
Comment (saghul) [inline ios/RCTWebRTC/ScreenCapturePickerViewManager.m:1]: Ditto.

File: ios/RCTWebRTC/WebRTCModule+Permissions.m [hunk 20]
```diff
@@ -23,13 +23,15 @@ - (AVMediaType)avMediaType:(NSString *)mediaType {
                   : (NSString *)mediaType resolver
                   : (RCTPromiseResolveBlock)resolve rejecter
                   : (RCTPromiseRejectBlock)reject) {
+#if TARGET_OS_TV
+    return;
```
Comment (saghul) [inline ios/RCTWebRTC/WebRTCModule+Permissions.m:27]: You need to resolve the promise before returning.

File: ios/RCTWebRTC/WebRTCModule+Permissions.m [hunk 21]
```diff
@@ -44,12 +46,16 @@ - (AVMediaType)avMediaType:(NSString *)mediaType {
...
     }
+#endif
 }
 
 RCT_EXPORT_METHOD(requestPermission
                   : (NSString *)mediaType resolver
                   : (RCTPromiseResolveBlock)resolve rejecter
                   : (RCTPromiseRejectBlock)reject) {
+#if TARGET_OS_TV
+    return;
```
Comment (saghul) [inline ios/RCTWebRTC/WebRTCModule+Permissions.m:57]: Same.

File: ios/RCTWebRTC/WebRTCModule+RTCMediaStream.m [hunk 22]
```diff
@@ -137,6 +148,10 @@ - (RTCVideoTrack *)createScreenCaptureVideoTrack {
 }
 
 RCT_EXPORT_METHOD(getDisplayMedia : (RCTPromiseResolveBlock)resolve rejecter : (RCTPromiseRejectBlock)reject) {
+#if TARGET_OS_TV
+    return;
```
Comment (saghul) [inline ios/RCTWebRTC/WebRTCModule+RTCMediaStream.m:152]: You need to reject the promise.

File: ios/RCTWebRTC/WebRTCModule+RTCMediaStream.m [hunk 23]
```diff
@@ -175,6 +191,9 @@ - (RTCVideoTrack *)createScreenCaptureVideoTrack {
                   : (NSDictionary *)constraints successCallback
                   : (RCTResponseSenderBlock)successCallback errorCallback
                   : (RCTResponseSenderBlock)errorCallback) {
+#if TARGET_OS_TV
+    return;
```
Comment (saghul) [inline ios/RCTWebRTC/WebRTCModule+RTCMediaStream.m:195]: Same.

File: ios/RCTWebRTC/WebRTCModule+RTCMediaStream.m [hunk 24]
```diff
@@ -234,11 +253,15 @@ - (RTCVideoTrack *)createScreenCaptureVideoTrack {
...
     self.localStreams[mediaStreamId] = mediaStream;
     successCallback(@[ mediaStreamId, tracks ]);
+#endif
 }
 
 #pragma mark - Other stream related APIs
 
 RCT_EXPORT_METHOD(enumerateDevices : (RCTResponseSenderBlock)callback) {
+#if TARGET_OS_TV
+    return;
```
Comment (saghul) [inline ios/RCTWebRTC/WebRTCModule+RTCMediaStream.m:263]: Call the callback with an empty array.

**Review by saghul — CHANGES_REQUESTED** [review]
A few more smal ones.

File: ios/RCTWebRTC/WebRTCModule+Permissions.m (hunk 20 above)
Comment (montanaeli) [reply ios/RCTWebRTC/WebRTCModule+Permissions.m:28]: This is the way you are meaning?

File: ios/RCTWebRTC/WebRTCModule+RTCMediaStream.m (hunk 23 above)
Comment (montanaeli) [reply ios/RCTWebRTC/WebRTCModule+RTCMediaStream.m:197]: I used errorCallback, is it okay?

**Comment by llspalex** [comment]
@montanaeli you mentioned that "local WebRTC.framework m112" is required. The [link](https://github.com/millicast/streaming-sdk-react-native-getting-started/blob/main/ios/libWebRTC.zip) you posted contains a WebRTC framework for tvOS - tvOS device and tvOS simulator binaries. But, the tvOS device binary seems incomplete, it only contains 4x symbols whereas the simulator binary contains 100+ symbols. 

Do you have an updated link for a WebRTC framework for tvOS (device) that contains all the symbols? 

**### tvOS simulator (full list not shown)**
Screenshot 2023-09-03 at 23 01 01

**### tvOS device (full list shown - only 4 symbols)**
Screenshot 2023-09-03 at 23 02 02

**Comment by montanaeli** [comment]
@llspalex I apologize for the inconvenience, we updated the WebRTC builds [here](https://github.com/millicast/streaming-sdk-react-native-getting-started/blob/feedback-fixes/ios/libWebRTC.zip)

File: ios/RCTWebRTC/WebRTCModule+Permissions.m (hunk 20 above)
Comment (saghul) [reply ios/RCTWebRTC/WebRTCModule+Permissions.m:28]: yep!

**Review by saghul — APPROVED** [review]

**Comment by saghul** [comment]
Cheers!

**Comment by 8BallBomBom** [comment]
Glad to see this drop before the next release 👍🏻 
Good work all around 😸

**Comment by montanaeli** [comment]
Thank you! 😄

**Comment by rkz98** [comment]
Does the data-channel protocol work now? has someone tested?

**Pull request merged**